R/reporter-progress.R

```diff
@@ -308,6 +338,113 @@ CompactProgressReporter <- R6::R6Class("CompactProgressReporter",
   )
 )
 
+# parallel progres reporter -----------------------------------------------
+
+#' @export
+#' @rdname ProgressReporter
+
+ParallelProgressReporter <- R6::R6Class("ParallelProgressReporter",
+  inherit = ProgressReporter,
+  public = list(
+
+    files = list(),
+    spin_frame = 0L,
+
+    initialize = function(...) {
+      super$initialize(...)
+      self$capabilities$parallel_support <- TRUE
+      self$capabilities$parallel_updates <- TRUE
+      self$update_interval <- 0.05
+    },
+
+    start_file = function(file)  {
+      if (! file %in% names(self$files)) {
+        self$files[[file]] <- list(
+          issues = Stack$new(),
+          n_fail = 0L,
+          n_skip_ = 0L,
+          n_warn = 0L,
+          n_ok = 0L,
+          name = context_name(file),
+          start_time = proc.time()
+        )
+      }
+      self$file_name <- file
+    },
+
+    start_context = function(context) {
+      # we'll just silently ignore this
+    },
+
+    end_context = function(context) {
+      # we'll just silently ignore this
+    },
+
+    end_file = function() {
+      fsts <- self$files[[self$file_name]]
+      time <- proc.time() - fsts$start_time
+
+      self$show_status(complete = TRUE, time = time[[3]], pad = TRUE)
+      self$report_issues(fsts$issues)
+
+      self$files[[self$file_name]] <- NULL
+      if (length(self$files)) self$update(force = TRUE)
+    },
+
+    end_reporter = function() {
+      self$cat_tight("\r", strpad(""))
+      super$end_reporter()
+    },
+
+    show_header = function() {
+      super$show_header()
+      self$update(force = TRUE)
+    },
+
+    status_data = function() {
+      self$files[[self$file_name]]
+    },
+
+    add_result = function(context, test, result) {
+      self$ctxt_n <- self$ctxt_n + 1L
+      file <- self$file_name
+      if (expectation_broken(result)) {
+        self$n_fail <- self$n_fail + 1
+        self$files[[file]]$n_fail <- self$files[[file]]$n_fail + 1L
+        self$files[[file]]$issues$push(result)
+      } else if (expectation_skip(result)) {
+        self$n_skip <- self$n_skip + 1
+        self$files[[file]]$n_skip <- self$files[[file]]$n_skip + 1L
+        self$files[[file]]$issues$push(result)
+        self$skips$push(result$message)
+      } else if (expectation_warning(result)) {
+        self$n_warn <- self$n_warn + 1
+        self$files[[file]]$n_warn <- self$files[[file]]$n_warn + 1L
+        self$files[[file]]$issues$push(result)
+      } else {
+        self$n_ok <- self$n_ok + 1
+        self$files[[file]]$n_ok <- self$files[[file]]$n_ok + 1
+      }
+    },
+
+    update = function(force = FALSE) {
+      if (!force && !self$should_update()) return()
+      self$spin_frame <- self$spin_frame + 1L
+      status <- spinner(self$frames, self$spin_frame)
+
+      message <- paste(
+        status,
+        summary_line(self$n_ok, self$n_fail, self$n_warn, self$n_skip),
+        if (length(self$files) > 0) "@" else " starting up...",
```

Comment:
```suggestion
        if (length(self$files) > 0) "@" else "Starting up...",
```

(no space needed since you use `paste()` on the next line)